compiler/rustc_ty_utils/src/abi.rs

```diff
@@ -784,6 +784,21 @@ fn fn_abi_adjust_for_abi<'tcx>(
                 // an LLVM aggregate type for this leads to bad optimizations,
                 // so we pick an appropriately sized integer type instead.
                 arg.cast_to(Reg { kind: RegKind::Integer, size });
+
+                // Let's see if we can add a `noundef`. This is only legal for arrays, definitely
+                // not for unions. This is also legal for `#[repr(transparent)] struct` or
+                // `#[repr(transparent)] enum` containing array. Note that `#[repr(transparent)]`
+                // can contain other `#[repr(transparent)]` structs or enums, which can eventually
+                // contain an array!
+                if arg.layout.ty.is_array() || is_transparent_array(cx, arg.layout) {
```

Comment:
I don't understand why arrays would be special here. What if this is an array of unions, or something like that?